src/Data/Constraint/Deriving/ToInstance.hs

```diff
@@ -152,8 +153,11 @@ toInstance (ToInstance omode) (NonRec bindVar bindExpr) = do
       Nothing -> pluginLocatedError loc notGoodMsg
       Just ma -> pure ma
     let matchedTy = substTyVar match varCls
-        instSig = mkSpecForAllTys bndrs $ mkFunTys theta matchedTy
-        bindBareTy = mkSpecForAllTys bndrs $ mkFunTys theta $ mkTyConApp tcBareConstraint [matchedTy]
+        instSig = mkSpecForAllTys bndrs $ mkVisFunTys theta matchedTy
+        bindBareTy = mkSpecForAllTys bndrs $ mkVisFunTys theta $ mkTyConApp tcBareConstraint [matchedTy]
+#if __GLASGOW_HASKELL__ < 810
+        mkVisFunTys = mkFunTys
+#endif
```

Comment:
I believe, here we should have  `mkInvisFunTys`, since we have constraint types on the left.